XdmodAID.md

```diff
@@ -0,0 +1,868 @@
+
+And finally for openstack, we need to effectively squash a layer of the hierarchy
+in order to match the jobs side, so pi2project just maps an openstack project
+to an openstack project that is found in the names.csv.
+
+The examples found in the xdmod documentation use abbreviations for unique keys. This
+doesn't suite an automated system as there are cases where there are multiple
+colleges/universities with the same name.  For example, I went to westminster college,
+at the time there were 5 westminster colleges.  Since, we really don't need a
+human readable ID, I found using numbers to be more convienent.
+
+After working throught that example, it is obvious that many of the hierarchy items
+need to have unique IDs.  Initially, I was under the impression that
+RegApp/ColdFront/KeyCloak would also need to have unique ids for this data, however,
+this was not the case.
+
+It was suggested that we use a cryptological signiture to give unique ids as the possiblity
+of a data collision is miniscule. The advantage here would be that there would be no
+database access to determine what the ID would be.  It turned out to be a bit more
+complicated, as each item in the hierarchy needed it's own unique.  It was simpler create
+a sequencer in the database and look up the keys when needed.  We needed a database
+anyways as we need to track the history of the hierarcy for potential audits and
+I am uncertain how long the information will be kept in coldfront or keycloak when
+users and projects are deleted.
+
+And here are the commands that load the CSVs into the xdmod database:
+```
+    xdmod-import-csv -t hierarchy -i hierarchy.csv
+    xdmod-import-csv -t group-to-hierarchy -i group.csv
+    xdmod-import-csv -t names -i names.csv
+    xdmod-import-csv -t cloud-project-to-pi -i pi2project.csv
+    acl-config
+```
+
+In termas of testing what I was originally attempting to do was to create a test
+container that was derived from the dockerfile that contained xdmod-hierarchy and
+included mariadb with all of the pytest moducles.  This could eventually be
+incorporated in to test sequence in github to run pytest each time the container
+was built on github as part of our CI process.
+
+This would also work with how I deploy xdmod on openshift using build
+configurations to test before actual deployment.  Furthermore, it could be used
+to test the production system insitu.
+
+As I was having difficulty with getting this to work I simplifed it a great
+deal, just to have the tests run inside a container.  In the process, found that
+for stand-a-lone installs, mariadb was creating root accounts without a password
+that had to be accessed from the root account on that system.  After trying
+several ways suggested by the mariadb documentation and from stack exchange,
+settled on using it the way it was designed.
+
+Another option would be to use a service container to host mariadb, however,
+this initially seemed a bit more complicated than just
```

Comment:
just what?

Reply:
I was probably falling asleep when I wrote that and dreamed that I had finished my thought.

I hope what I wrote is better.